### PR TITLE
docs(framework) Document public/private API approach

### PR DESCRIPTION
**Pull request by danieljanes**

### CODE REVIEW

File: doc/source/contributor-explanation-public-and-private-apis.rst [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+What does this mean? It means that ``client``, ``common``, ``server`` and ``simulation`` are part of the public API, but ``cli`` and ``proto`` are not.
+The ``flwr`` subpackages ``cli`` and ``proto`` are private APIs.
+A private API can change completely from one release to the next (even in patch releases).
+It can change in a breaking way, it can be renamed (for example, ``flwr.cli`` could be renamed to ``flwr.command``) and it can even be removed completely.
+
+Therefore, as a Flower user:
+
+- ``from flwr import client`` ✅ Ok, you're importing a public API.
+- ``from flwr import proto`` ❌ Not ok, you're importing a private API.
```
Comment (danieljanes) [inline doc/source/contributor-explanation-public-and-private-apis.rst:63]:
```suggestion
- ``from flwr import proto`` ❌ Not recommended, you're importing a private API.
```

File: doc/source/contributor-explanation-public-and-private-apis.rst [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+What's notable here is that all strategies are implemented in dedicated modules (e.g., ``fedavg.py``).
+In ``__init__.py``, we *import* the components we want to make part of the public API and then *export* them via ``__all__``.
+Note that we export the component itself (for example, the ``FedAvg`` class), but not the module it is defined in (for example, ``fedavg.py``).
+This allows us to move the definition of ``FedAvg`` into a different module (or even a module in a subpackage) without breaking the public API (as long as we update the import path in ``__init__.py``).
+
+Therefore:
+
+- ``from flwr.server.strategy import FedAvg`` ✅ Ok, you're importing a class that is part of the public API.
+- ``from flwr.server.strategy import fedavg`` ❌ Not ok, you're importing a private module.
```
Comment (danieljanes) [inline doc/source/contributor-explanation-public-and-private-apis.rst:87]:
```suggestion
- ``from flwr.server.strategy import fedavg`` ❌ Not recommended, you're importing a private module.
```

File: doc/source/contributor-explanation-public-and-private-apis.rst [hunk 3]
```diff
@@ -0,0 +1,118 @@
+Public and private APIs
+=======================
+
+In Python, everything is public.
+To enable developers to understand which components can be relied upcon, Flower declares a public API.
```
Comment (tanertopal) [inline doc/source/contributor-explanation-public-and-private-apis.rst:5]: I think there is a typo: relied upon

Comment (danieljanes) [reply doc/source/contributor-explanation-public-and-private-apis.rst:5]:
```suggestion
To enable developers to understand which components can be relied upon, Flower declares a public API.
```

**Review by tanertopal — APPROVED** [review]

**Pull request merged**